### PR TITLE
Python3 doctest compatibility: algorithms SmoothData to TimeSlice

### LINKED ISSUE #20867: Python 3 doctest compatibility smoothdata to timeslice
I've taken 
```
SmoothData-v1
SmoothMD-v1
SmoothNeighbours-v1
SofQW-v1
SofQWCentre-v1
SofQWMoments-v1
SofQWMomentsScan-v1
SofQWNormalisedPolygon-v1
SofQWPolygon-v1
SolidAngle-v1
SortByQVectors-v1
SortDetectors-v1
SortEvents-v1
SortHKL-v1
SortPeaksWorkspace-v1
SortTableWorkspace-v1
SpatialGrouping-v1
SpecularReflectionCalculateTheta-v1
SpecularReflectionCalculateTheta-v2
SpecularReflectionPositionCorrect-v1
SpecularReflectionPositionCorrect-v2
SphericalAbsorption-v1
SplineBackground-v1
SplineInterpolation-v1
SplineSmoothing-v1
Squares-v1
StartLiveData-v1
StatisticsOfPeaksWorkspace-v1
StatisticsOfTableWorkspace-v1
Stitch1D-v3
Stitch1DMany-v1
StringToPng-v1
StripVanadiumPeaks-v1
StripVanadiumPeaks-v2
SuggestTibCNCS-v1
SuggestTibHYSPEC-v1
SumEventsByLogValue-v1
SumNeighbours-v1
SumRowColumn-v1
SumSpectra-v1
SwapWidths-v1
Symmetrise-v1
TOFTOFCropWorkspace-v1
TOFTOFMergeRuns-v1
TOSCABankCorrection-v1
ThresholdMD-v1
TimeSlice-v1
```

### PR DESCRIPTION
**Pull request by NickDraper**
Python3 doctest compatibility: algorithms  SmoothData to TimeSlice

## Description of work.
This PR is part of refactoring effort to bring Python3 compatiblity into the doctests.
See https://github.com/mantidproject/mantid/issues/20672 for details.

The following tests were refactored to be compatible :
```
SmoothData-v1
SmoothMD-v1
SmoothNeighbours-v1
SofQW-v1
SofQWCentre-v1
SofQWMoments-v1
SofQWMomentsScan-v1
SofQWNormalisedPolygon-v1
SofQWPolygon-v1
SolidAngle-v1
SortByQVectors-v1
SortDetectors-v1
SortEvents-v1
SortHKL-v1
SortPeaksWorkspace-v1
SortTableWorkspace-v1
SpatialGrouping-v1
SpecularReflectionCalculateTheta-v1
SpecularReflectionCalculateTheta-v2
SpecularReflectionPositionCorrect-v1
SpecularReflectionPositionCorrect-v2
SphericalAbsorption-v1
SplineBackground-v1
SplineInterpolation-v1
SplineSmoothing-v1
Squares-v1
StartLiveData-v1
StatisticsOfPeaksWorkspace-v1
StatisticsOfTableWorkspace-v1
Stitch1D-v3
Stitch1DMany-v1
StringToPng-v1
StripVanadiumPeaks-v1
StripVanadiumPeaks-v2
SuggestTibCNCS-v1
SuggestTibHYSPEC-v1
SumEventsByLogValue-v1
SumNeighbours-v1
SumRowColumn-v1
SumSpectra-v1
SwapWidths-v1
Symmetrise-v1
TOFTOFCropWorkspace-v1
TOFTOFMergeRuns-v1
TOSCABankCorrection-v1
ThresholdMD-v1
TimeSlice-v1
```

## To test.

Code review.

Fixes #20867
Internal change, no release notes

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/SortEvents-v1.rst [hunk 1]
```diff
@@ -33,7 +33,7 @@ Usage
   ptimes = ev1.getPulseTimes()
   tofs = ev1.getTofs()
   for eindex in xrange(10):
```
Comment (rosswhitfield) [inline docs/source/algorithms/SortEvents-v1.rst]: Use `range` not `xrange`

File: docs/source/algorithms/SortEvents-v1.rst [hunk 2]
```diff
@@ -66,7 +66,7 @@ Output:
...
```
Comment (rosswhitfield) [inline docs/source/algorithms/SortEvents-v1.rst]: `xrange->range`

File: docs/source/algorithms/SortTableWorkspace-v1.rst [hunk 3]
```diff
@@ -42,15 +42,15 @@ Usage
 
     # Sort in ascending order
     sorted_asc = SortTableWorkspace(table, Columns = ['x','y','z'])
-    print 'Sorted ascending'
+    print('Sorted ascending')
     for i in range(sorted_asc.rowCount()):
-        print sorted_asc.row(i)
+        print(sorted_asc.row(i))
```
Comment (rosswhitfield) [inline docs/source/algorithms/SortTableWorkspace-v1.rst]: These tests are printing a dict which you then can't assume the order of the columns for python3
```
Expected:
    Sorted ascending
    {'y': 'one (1)', 'x': 1, 'z': 7.0}
    {'y': 'one (2)', 'x': 1, 'z': 2.0}
    {'y': 'one (3)', 'x': 1, 'z': 1.0}
    {'y': 'two (1)', 'x': 2, 'z': 3.0}
    {'y': 'two (1)', 'x': 2, 'z': 8.0}
    {'y': 'two (2)', 'x': 2, 'z': 6.0}
    {'y': 'two (2)', 'x': 2, 'z': 9.0}
    {'y': 'three (2)', 'x': 3, 'z': 4.0}
    {'y': 'three (2)', 'x': 3, 'z': 5.0}
    {'y': 'three (3)', 'x': 3, 'z': 0.0}
    Sorted descending
    {'y': 'three (3)', 'x': 3, 'z': 0.0}
    {'y': 'three (2)', 'x': 3, 'z': 5.0}
    {'y': 'three (2)', 'x': 3, 'z': 4.0}
    {'y': 'two (2)', 'x': 2, 'z': 9.0}
    {'y': 'two (2)', 'x': 2, 'z': 6.0}
    {'y': 'two (1)', 'x': 2, 'z': 8.0}
    {'y': 'two (1)', 'x': 2, 'z': 3.0}
    {'y': 'one (3)', 'x': 1, 'z': 1.0}
    {'y': 'one (2)', 'x': 1, 'z': 2.0}
    {'y': 'one (1)', 'x': 1, 'z': 7.0}
Got:
    Sorted ascending
    {'x': 1, 'y': 'one (1)', 'z': 7.0}
    {'x': 1, 'y': 'one (2)', 'z': 2.0}
    {'x': 1, 'y': 'one (3)', 'z': 1.0}
    {'x': 2, 'y': 'two (1)', 'z': 3.0}
    {'x': 2, 'y': 'two (1)', 'z': 8.0}
    {'x': 2, 'y': 'two (2)', 'z': 6.0}
    {'x': 2, 'y': 'two (2)', 'z': 9.0}
    {'x': 3, 'y': 'three (2)', 'z': 4.0}
    {'x': 3, 'y': 'three (2)', 'z': 5.0}
    {'x': 3, 'y': 'three (3)', 'z': 0.0}
    Sorted descending
    {'x': 3, 'y': 'three (3)', 'z': 0.0}
    {'x': 3, 'y': 'three (2)', 'z': 5.0}
    {'x': 3, 'y': 'three (2)', 'z': 4.0}
    {'x': 2, 'y': 'two (2)', 'z': 9.0}
    {'x': 2, 'y': 'two (2)', 'z': 6.0}
    {'x': 2, 'y': 'two (1)', 'z': 8.0}
    {'x': 2, 'y': 'two (1)', 'z': 3.0}
    {'x': 1, 'y': 'one (3)', 'z': 1.0}
    {'x': 1, 'y': 'one (2)', 'z': 2.0}
    {'x': 1, 'y': 'one (1)', 'z': 7.0}
```

File: docs/source/algorithms/SortDetectors-v1.rst [hunk 4]
```diff
@@ -22,11 +22,11 @@ Usage
...
    # No upstream detectors
-   print "Type of upID:",type(upIDs)
-   print "Number of upDistances:",upDistances.shape[0]
+   print("Type of upID: {}".format(type(upIDs)))
+   print("Number of upDistances: {}".format(upDistances.shape[0]))
    #Downstream detectors
-   print "First few values of downIDs: ", downIDs[0:5]
-   print "First few values of downDistances: ", downDistances[0], downDistances[1], downDistances[2], downDistances[3],downDistances[4]
+   print("First few values of downIDs: {}".format(downIDs[0:5]))
+   print("First few values of downDistances: {} {} {} {} {}".format(downDistances[0], downDistances[1], downDistances[2], downDistances[3],downDistances[4]))
```
Comment (rosswhitfield) [inline docs/source/algorithms/SortDetectors-v1.rst]: I see different precision with python 3
```
Expected:
    First few values of downDistances: 5.0 5.0000064 5.0000064 5.00001279998 5.00002559993
Got:
    First few values of downDistances: 5.0 5.000006399995904 5.000006399995904 5.000012799983616 5.000025599934465
```

File: docs/source/algorithms/SplineBackground-v1.rst [hunk 5]
```diff
@@ -36,13 +36,13 @@ Usage
...
   # Sum the squares of the differences of elements of arrays y and yout
-  print 'Fit quality is',np.sum( (y - yout)**2 )
+  print('Fit quality is {}'.format(np.sum( (y - yout)**2 )))
 
   # Increase the number of spline coefficients
   out = SplineBackground(ws, WorkspaceIndex=0, NCoeff=20)
   yout = out.readY(0)
   # Sum the squares of the differences of elements of arrays y and yout
-  print 'Fit quality is',np.sum( (y - yout)**2 )
+  print('Fit quality is {}'.format(np.sum( (y - yout)**2 )))
```
Comment (rosswhitfield) [inline docs/source/algorithms/SplineBackground-v1.rst]: I see a different precision with python 3
```
Expected:
    Fit quality is 0.0253975992834
    Fit quality is 1.53267607188e-07
Got:
    Fit quality is 0.025397599283365133
    Fit quality is 1.5326760718776523e-07
```

File: docs/source/algorithms/StartLiveData-v1.rst [hunk 6]
```diff
@@ -158,7 +158,7 @@ Usage
 
     #get the ouput workspace
     wsOut = mtd["wsOut"]
-    print "The workspace contains %i events" % wsOut.getNumberEvents()
+    print("The workspace contains %i events" % wsOut.getNumberEvents())
```
Comment (rosswhitfield) [inline docs/source/algorithms/StartLiveData-v1.rst]: Python 3 fails on the `except Exception, exc:` above

File: docs/source/algorithms/StatisticsOfTableWorkspace-v1.rst [hunk 7]
```diff
@@ -32,7 +32,7 @@ Usage
     for idx in range(stats.rowCount()):
         stat_name = stats.column('statistic')[idx]
         stat_value = stats.column('a')[idx]
-        print '%s of column \'a\' is %.3f' % (stat_name, stat_value)
+        print('%s of column \'a\' is %.3f' % (stat_name, stat_value))
```
Comment (rosswhitfield) [inline docs/source/algorithms/StatisticsOfTableWorkspace-v1.rst]: printing a dict can give different order in python 3
```
Expected:
    standard_deviation of column 'a' is 1.118
    minimum of column 'a' is 1.000
    median of column 'a' is 2.500
    maximum of column 'a' is 4.000
    mean of column 'a' is 2.500
Got:
    standard_deviation of column 'a' is 1.118
    maximum of column 'a' is 4.000
    minimum of column 'a' is 1.000
    mean of column 'a' is 2.500
    median of column 'a' is 2.500
```

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
There are also some `print` statements in `TOSCABankCorrection,TOFTOFCropWorkspace,TOFTOFMergeRuns` that need to fixed.

Edit: You appear to have changed these files in #20870 instead. Just update the descriptions of what docs you have changed.

File: docs/source/algorithms/StatisticsOfTableWorkspace-v1.rst (hunk 7 above)
Comment (NickDraper) [reply docs/source/algorithms/StatisticsOfTableWorkspace-v1.rst]: Interesting, this is a problem with the algorithm not the test.  Still I'll fix it anyway.

File: docs/source/algorithms/SortDetectors-v1.rst [hunk 8]
```diff
@@ -22,11 +22,11 @@ Usage
    upIDs,upDistances,downIDs,downDistances=SortDetectors(ws)
    
    # No upstream detectors
-   print "Type of upID:",type(upIDs)
-   print "Number of upDistances:",upDistances.shape[0]
+   print("Type of upID: {}".format(type(upIDs)))
```
Comment (rosswhitfield) [inline docs/source/algorithms/SortDetectors-v1.rst]: I didn't realize before but this is also different for python 3
```
Expected:
    Type of upID: <type 'numpy.ndarray'>
Got:
    Type of upID: <class 'numpy.ndarray'>
```

**Review by rosswhitfield — CHANGES_REQUESTED** [review]

File: docs/source/algorithms/SortDetectors-v1.rst (hunk 8 above)
Comment (NickDraper) [reply docs/source/algorithms/SortDetectors-v1.rst]: I'll put another ellipsis in there, that should handle it.

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
There is another `except Exception, exc:` in `StartLiveData`.

`StatisticsOfTableWorkspace` still fails with the same issue
```
Expected:
    standard_deviation of column 'a' is 1.118
    minimum of column 'a' is 1.000
    median of column 'a' is 2.500
    maximum of column 'a' is 4.000
    mean of column 'a' is 2.500
Got:
    standard_deviation of column 'a' is 1.118
    maximum of column 'a' is 4.000
    minimum of column 'a' is 1.000
    mean of column 'a' is 2.500
    median of column 'a' is 2.500
```

**Comment by NickDraper** [comment]
@rosswhitfield Do you have a python 3 build that can run Mantidplot?

If so can you run this script, take a look at the order in the Table workspace itself and let me know what it is?

I'm trying to separate out if this is due to the filling of the Table workspace or the reading from it.

**Comment by NickDraper** [comment]
This is with Python 2:
image

**Comment by rosswhitfield** [comment]
table

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
`StatisticsOfTableWorkspace` still fails for the same reason. I don't think the solution is the change the algorithm as it does the correct thing but just returns the results in a different order.  How about just changing the test to define the order?
```diff
     stats = StatisticsOfTableWorkspace(ws)
 
-    for idx in range(stats.rowCount()):
-        stat_name = stats.column('statistic')[idx]
+    for stat_name in ['standard_deviation','minimum', 'median', 'maximum','mean']:
+       idx = stats.column('statistic').index(stat_name)
         stat_value = stats.column('a')[idx]
         print('%s of column \'a\' is %.3f' % (stat_name, stat_value))
```

There are some `print` statements remaining in `StartLiveData` that need to be fixed.

**Review by rosswhitfield — APPROVED** [review]

**Pull request merged**